Feature/#4 add flag for config file

ndoering opened the pull request:
These commits implement the search for an commandline argument, that indicates a specific configuration file.

This closes issue #4

Commit:
Add command-line flag -config_file

To specify a config file at startup, this adds a routine to check for a
command-line flag. The flag, which is used is config_flag which is to be
followed by a string, that designates the config file. If multiple of
these flags a given, only the first is used. If no flag is specified,
the atom no_path is used. The string is given to tep_sup:start_link/1
for further use.

Commit:
Only take one config_file_path if multiple given

Commit:
Merge branches 'develop' and 'feature/#4-add-flag-for-config-file' of github.com:j-frost/tep into feature/#4-add-flag-for-config-file

Commit:
Allow config_file flag to be passed to tep

In order to change the config file used by TEP, the tep_app now searches
for a config_file argument and passes it through the tep_sup supervisor
to the configuration_server. Therefore the init/1 function needed to be
changed to take either the path to the config file or the atom no_path,
which is indicating, that no path was given to the Erlang VM.

j-frost commented on src/configuration_server.erl:
```diff
@@ -68,13 +68,20 @@ get_config_file_path() ->
-    ConfigFilePath = search_for_config_file_path(
+init([no_path]) ->
+    ConfigFilePathResult = search_for_config_file_path(
         ?TEP_DEFAULT_CONFIG_LOCATIONS,
         ?TEP_DEFAULT_CONFIG_FILE_NAMES
     ),
-    Configuration = create_configuration(ConfigFilePath),
-    return_from_init(Configuration).
+    ConfigurationResult = create_configuration(ConfigFilePathResult),
+    return_from_init(ConfigurationResult);
+init([ConfigFilePath]) ->
+    ConfigFilePathResult = search_for_config_file_path(
+        ["./"],
+        [ConfigFilePath]
+    ),
+    ConfigurationResult = create_configuration(ConfigFilePathResult),
+    return_from_init(ConfigurationResult).
```
Hoo boy, my TypeScript eyes just wanna
```typescript
export function init(configFilePath: string | undefined) {
  const configFilePathResult = searchForConfigFilePath(
    configFilePath ? 
      { searchDir: './', fileName: configFilePath } :
      { searchDir: DEFAULT/* stuff */, ... }
  );
  // ...
}
```
soooo bad. I know what we have is probably "right" in the sense that it's Erlang idiomatic, but I'm still getting used to what that means, honestly...

ndoering replied:
I don't know why we created this repeating mess, but I have implemented a simpler and cleaner version.

j-frost commented on test/configuration_server_test.erl:
```diff
@@ -52,11 +52,38 @@ init_test() ->
+        (Path) -> meck:passthrough([Path])
+    end),
+    ?assertEqual(
+        {ok, {configuration, "./tep_test.ini"}},
```
So this is not an issue currently, but just a thought: If our `configuration` record changes (and presumably gains more fields), will that destroy these tests every time we change it? Can we easily test only parts of tuples for equality? In case we can't, maybe we should start thinking about whether that influences our design in any way.

ndoering replied:
Since the configuration is a central part of the application and we might use versioning and hot loading code at some time in the future, I would recommend treating this with care and check every change to it.

j-frost commented on test/configuration_server_test.erl:
```diff
@@ -52,11 +52,38 @@ init_test() ->
+        (Path) -> meck:passthrough([Path])
+    end),
+    ?assertEqual(
+        {stop, {no_file, "No config file found"}},
```
So this is a nit pick, but as a user I'd appreciate different error messages for "no file found" (because you specified none and I had to go looking for one) and "the file you specified doesn't exist".

j-frost commented on test/tep_app_test.erl:
```diff
@@ -0,0 +1,10 @@
+-include_lib("eunit/include/eunit.hrl").
+
+get_config_argument_valid_test() ->
+    Valid = {ok, [["test.ini", "config.ini"], ["tep.ini"]]},
```
I don't understand these `Args`. Is that just a quirk of the `application` behaviour's `init:get_argument` function? In either case, do we document that as a comment or something?

j-frost replied:
Hm ok I just looked up the behaviour of `init:get_argument`. I would have expected tep.ini to be the argument that gets passed to the actual application, because it's the last and "overrides" the others. And also because it doesn't really make sense to pass multiple arguments to the "config file" flag. In fact I'd advocate for us to refuse startup when given multiple config files, but that's pedantic and I'm prepared to drop it. 
Back to the test: The way it's written now is a little obtuse, but I don't know how to well fix it. Could link to docs, but that can sometimes go out of date much faster than anticipated...

ndoering replied:
Since merging of a set of configuration files is a path to hell, I would recommend, not to try this. I think its equal to take the first or the last value. We might discuss whether we stop the startup completely, if multiple `config_file` arguments are found. I'm indifferent about this.

j-frost commented on src/tep_app.erl:
```diff
@@ -6,10 +6,27 @@
 %% internal functions
+-spec get_config_file_path() ->
+    string() | no_path.
+get_config_file_path() ->
```
Missing docstring

ndoering replied:
Will be added

j-frost commented on src/tep_app.erl:
```diff
@@ -6,10 +6,27 @@
+
+-spec get_config_argument({ok, Args :: [Values :: [string()]]} | error) ->
+    string() | no_path.
+get_config_argument({ok, [[Value | _] | _]}) -> Value;
```
Missing docstring

ndoering replied:
Will be added

j-frost commented on src/tep_app.erl:
```diff
@@ -6,10 +6,27 @@
 -export([start/2, stop/1]).
 
-start(_StartType, _StartArgs) -> tep_sup:start_link().
+start(_StartType, _StartArgs) -> tep_sup:start_link(get_config_file_path()).
```
That **doesn't** get formatted onto two lines? Kinda weird imho... I don't yet understand this formatter's preferences.

ndoering replied:
Same here

j-frost commented on test/tep_app_test.erl:
```diff
@@ -0,0 +1,10 @@
+    ?assertEqual("test.ini", tep_app:get_config_argument(Valid)).
+
+get_config_argument_invalid_test() ->
+    ?assertEqual(no_path, tep_app:get_config_argument(error)).
```
Sounds a little severe to me. Maybe something like `no_custom_config_path`?

ndoering replied:
This is a good name, I have change it.

j-frost commented on src/tep_sup.erl:
```diff
@@ -6,14 +6,21 @@
 
-start_link() ->
-    supervisor:start_link({local, ?SERVER}, ?MODULE, []).
+-spec start_link(string() | no_path) ->
+    {ok, pid()} |
+    ignore |
+    {error,
+        {already_started, pid()} |
+        {shutdown, term()} |
+        term()}.
```
Wait, we didn't have this? Did we miss enforcing `spec`s on every function? Because I kinda wanna do that now.

ndoering replied:
Same here. I think it should be necessary. Please create an issue fo that.

j-frost commented on src/tep_sup.erl:
```diff
@@ -6,14 +6,21 @@
+        {shutdown, term()} |
+        term()}.
+start_link(ConfigPath) ->
+    supervisor:start_link({local, ?SERVER}, ?MODULE, [ConfigPath]).
```
It's actually possible to error with "already started". Seems like a good idea for our root supervisor, no?

ndoering replied:
This is implicitly used by the calling supervisor, IRC. No need to implement it by hand, I presume.

j-frost commented on src/tep_sup.erl:
```diff
@@ -24,14 +31,20 @@ start_link() ->
 %%                  modules => modules()}   % optional
-init([]) ->
+init([ConfigPath]) ->
     SupFlags = #{strategy => one_for_all, intensity => 0, period => 1},
+
+    Configuration = #{
+        id => configuration,
+        start => {configuration_server, start_link, [ConfigPath]}
+    },
+
     Storage = #{
         id => storage_sup,
         start => {storage_sup, start_link, []},
         shutdown => infinity,
         type => supervisor
     },
-    {ok, {SupFlags, [Storage]}}.
+    {ok, {SupFlags, [Configuration, Storage]}}.
```
Can't find the tests that require this behaviour? Can we not reasonably test this?

ndoering replied:
Since this would require changing the test every time we change something here, this is the same as the configuration file that you remarked above. Also, there is absolutely no logic in this. Whats your opinion on this?

Commit:
Refactor init/1 function of configuration_server

The init/1 function of the configuration_server module was repetitive
and unnecessarily complex. This is streamlined with this commit.
